[PyTorch] Use intrusive_ptr(p, DontIncreaseRefcount) directly in TensorBase unsafe borrow ctor

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138934

We observed ASAN failures stemming from https://github.com/pytorch/pytorch/blob/5ea67778619c31b13644914deef709199052ee55/torch/csrc/autograd/python_variable.cpp#L403 . Since it's possible that `tensor` is dead here, `borrowed()` needs to avoid dereferencing it. `intrusive_ptr::reclaim` dereferences the pointer in builds with debug checks enabled, so use the DontIncreaseRefcount ctor directly instead.

Differential Revision: [D64990707](https://our.internmc.facebook.com/intern/diff/D64990707/)

swolchok committed:
[PyTorch] Use intrusive_ptr(p, DontIncreaseRefcount) directly in TensorBase unsafe borrow ctor

We observed ASAN failures stemming from https://github.com/pytorch/pytorch/blob/5ea67778619c31b13644914deef709199052ee55/torch/csrc/autograd/python_variable.cpp#L403 . Since it's possible that `tensor` is dead here, `borrowed()` needs to avoid dereferencing it. `intrusive_ptr::reclaim` dereferences the pointer in builds with debug checks enabled, so use the DontIncreaseRefcount ctor directly instead.

Differential Revision: [D64990707](https://our.internmc.facebook.com/intern/diff/D64990707/)

[ghstack-poisoned]

Skylion007 commented on aten/src/ATen/core/TensorBase.h:
```diff
@@ -88,7 +88,7 @@ class TORCH_API TensorBase {
   // time. Intended to support MaybeOwnedTraits<Tensor>.
   explicit TensorBase(unsafe_borrow_t, const TensorBase& rhs)
-      : impl_(c10::intrusive_ptr<at::TensorImpl, UndefinedTensorImpl>::reclaim(rhs.impl_.get())) {}
+      : impl_(c10::intrusive_ptr<at::TensorImpl, UndefinedTensorImpl>(rhs.impl_.get(), c10::raw::DontIncreaseRefcount{})) {}
```
@cyyever This doesn't resolve the slicing issues in your PR correct?

cyyever replied:
@Skylion007 May not be the same issue, but here some tests fail.

cyyever commented:
@pytorchbot rebase -b main

swolchok committed:
Update

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge

swolchok committed:
Update on "[PyTorch] Use intrusive_ptr(p, DontIncreaseRefcount) directly in TensorBase unsafe borrow ctor"


We observed ASAN failures stemming from https://github.com/pytorch/pytorch/blob/5ea67778619c31b13644914deef709199052ee55/torch/csrc/autograd/python_variable.cpp#L403 . Since it's possible that `tensor` is dead here, `borrowed()` needs to avoid dereferencing it. `intrusive_ptr::reclaim` dereferences the pointer in builds with debug checks enabled, so use the DontIncreaseRefcount ctor directly instead.

Differential Revision: [D64990707](https://our.internmc.facebook.com/intern/diff/D64990707/)

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge

cyyever commented:
@pytorchmergebot merge -f "Unrelated failures."
